# 1. Data Exploration and Preprocessing


In [30]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [31]:
data = pd.read_csv('adult_with_headers.csv')
data.head()

age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [32]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


**OUR DATA HAS 15 COLUMNS, OUT OF WHICH,**

*   **9 ARE STRING**
*   **6 ARE INTERGER**



In [33]:
categorical_data = data.select_dtypes(include=['object'])
categorical_data

workclass    education       marital_status  \
0              State-gov    Bachelors        Never-married   
1       Self-emp-not-inc    Bachelors   Married-civ-spouse   
2                Private      HS-grad             Divorced   
3                Private         11th   Married-civ-spouse   
4                Private    Bachelors   Married-civ-spouse   
...                  ...          ...                  ...   
32556            Private   Assoc-acdm   Married-civ-spouse   
32557            Private      HS-grad   Married-civ-spouse   
32558            Private      HS-grad              Widowed   
32559            Private      HS-grad        Never-married   
32560       Self-emp-inc      HS-grad   Married-civ-spouse   

               occupation    relationship    race      sex  native_country  \
0            Adm-clerical   Not-in-family   White     Male   United-States   
1         Exec-managerial         Husband   White     Male   United-States   
2       Handlers-cleaners   Not-in-family   White     Male   United-States   
3       Handlers-cleaners         Husband   Black     Male   United-States   
4          Prof-specialty            Wife   Black   Female            Cuba   
...                   ...             ...     ...      ...             ...   
32556        Tech-support            Wife   White   Female   United-States   
32557   Machine-op-inspct         Husband   White     Male   United-States   
32558        Adm-clerical       Unmarried   White   Female   United-States   
32559        Adm-clerical       Own-child   White     Male   United-States   
32560     Exec-managerial            Wife   White   Female   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 9 columns]

In [34]:
for i in categorical_data.columns:
    count = data[i].value_counts()
    print(count)

 Private             22696
 Self-emp-not-inc     2541
 Local-gov            2093
 ?                    1836
 State-gov            1298
 Self-emp-inc         1116
 Federal-gov           960
 Without-pay            14
 Never-worked            7
Name: workclass, dtype: int64
 HS-grad         10501
 Some-college     7291
 Bachelors        5355
 Masters          1723
 Assoc-voc        1382
 11th             1175
 Assoc-acdm       1067
 10th              933
 7th-8th           646
 Prof-school       576
 9th               514
 12th              433
 Doctorate         413
 5th-6th           333
 1st-4th           168
 Preschool          51
Name: education, dtype: int64
 Married-civ-spouse       14976
 Never-married            10683
 Divorced                  4443
 Separated                 1025
 Widowed                    993
 Married-spouse-absent      418
 Married-AF-spouse           23
Name: marital_status, dtype: int64
 Prof-specialty       4140
 Craft-repair         4099
 Exec-managerial

**The missing values are saved as '?', lets replace it and then treat it.**

**Some columns have '?' has values, these are missing values. So lets replace them first.**

In [35]:
# Replace '?' with NaN
data.replace('?', pd.NA, inplace=True)

In [36]:
data.dropna(inplace=True)

In [37]:
for i in categorical_data.columns:
    count = data[i].value_counts()
    print(count)

 Private             22696
 Self-emp-not-inc     2541
 Local-gov            2093
 ?                    1836
 State-gov            1298
 Self-emp-inc         1116
 Federal-gov           960
 Without-pay            14
 Never-worked            7
Name: workclass, dtype: int64
 HS-grad         10501
 Some-college     7291
 Bachelors        5355
 Masters          1723
 Assoc-voc        1382
 11th             1175
 Assoc-acdm       1067
 10th              933
 7th-8th           646
 Prof-school       576
 9th               514
 12th              433
 Doctorate         413
 5th-6th           333
 1st-4th           168
 Preschool          51
Name: education, dtype: int64
 Married-civ-spouse       14976
 Never-married            10683
 Divorced                  4443
 Separated                 1025
 Widowed                    993
 Married-spouse-absent      418
 Married-AF-spouse           23
Name: marital_status, dtype: int64
 Prof-specialty       4140
 Craft-repair         4099
 Exec-managerial

In [38]:
numerical_data = data.select_dtypes(include=['int64'])
numerical_data

age  fnlwgt  education_num  capital_gain  capital_loss  hours_per_week
0       39   77516             13          2174             0              40
1       50   83311             13             0             0              13
2       38  215646              9             0             0              40
3       53  234721              7             0             0              40
4       28  338409             13             0             0              40
...    ...     ...            ...           ...           ...             ...
32556   27  257302             12             0             0              38
32557   40  154374              9             0             0              40
32558   58  151910              9             0             0              40
32559   22  201490              9             0             0              20
32560   52  287927              9         15024             0              40

[32561 rows x 6 columns]

**STANDARD SCALING**

In [39]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
numerical_data = scaler.fit_transform(numerical_data)

In [40]:

numerical_data = pd.DataFrame(numerical_data, columns=data.select_dtypes(include=['int64']).columns)


***1. Standard Scaling (Z-Score Normalization)***

Standard scaling transforms the data so that it has a mean of 0 and a standard deviation of 1. This is done by subtracting the mean and dividing by the standard deviation for each feature.

***When to Use:***

1. Standard scaling is preferred when the algorithms assume normally distributed data or are sensitive to the magnitude of features.

***Examples include:***
* Linear Regression: Assumes that the predictors are centered around 0.

* Logistic Regression: Assumes that the predictors are normally distributed.

* Principal Component Analysis (PCA): Maximizes variance, so scaling features with different variances equally is important.

* K-Nearest Neighbors (KNN): Distance-based methods are sensitive to the magnitude of features.

* Support Vector Machines (SVM): The distance between data points is crucial, and unscaled features can distort this.
Data with Outliers: Standard scaling is more robust to outliers compared to Min-Max scaling, as it centers the data around the mean and does not compress the range between min and max values.


**2. Min-Max Scaling (Normalization)**

Definition: Min-Max scaling transforms the data to fit within a specific range, usually [0, 1]. This is done by subtracting the minimum value and dividing by the range of the feature.



***When to Use:***
1. Min-Max scaling is preferred when algorithms assume that all features have the same range or are bounded.

***Examples include:***

* Neural Networks: Activation functions like sigmoid and tanh are sensitive to the range of input features. Min-Max scaling helps in converging faster during training.

* K-Means Clustering: Distance-based methods benefit from normalized feature ranges to ensure equal weighting.

* Algorithms with Bounds: Algorithms that involve bounded functions (e.g., gradient descent) benefit from normalized features to ensure the stability and performance of the model.

* Data without Outliers: Min-Max scaling is more sensitive to outliers because the minimum and maximum values are directly used. If the data has significant outliers, Min-Max scaling might not be appropriate as it will compress the range of the majority of data points.

# Encoding Techniques

In [41]:
categorical_data

workclass    education       marital_status  \
0              State-gov    Bachelors        Never-married   
1       Self-emp-not-inc    Bachelors   Married-civ-spouse   
2                Private      HS-grad             Divorced   
3                Private         11th   Married-civ-spouse   
4                Private    Bachelors   Married-civ-spouse   
...                  ...          ...                  ...   
32556            Private   Assoc-acdm   Married-civ-spouse   
32557            Private      HS-grad   Married-civ-spouse   
32558            Private      HS-grad              Widowed   
32559            Private      HS-grad        Never-married   
32560       Self-emp-inc      HS-grad   Married-civ-spouse   

               occupation    relationship    race      sex  native_country  \
0            Adm-clerical   Not-in-family   White     Male   United-States   
1         Exec-managerial         Husband   White     Male   United-States   
2       Handlers-cleaners   Not-in-family   White     Male   United-States   
3       Handlers-cleaners         Husband   Black     Male   United-States   
4          Prof-specialty            Wife   Black   Female            Cuba   
...                   ...             ...     ...      ...             ...   
32556        Tech-support            Wife   White   Female   United-States   
32557   Machine-op-inspct         Husband   White     Male   United-States   
32558        Adm-clerical       Unmarried   White   Female   United-States   
32559        Adm-clerical       Own-child   White     Male   United-States   
32560     Exec-managerial            Wife   White   Female   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 9 columns]

**Label Encoding**

In [42]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

for col in ['race', 'sex', 'income']:
    categorical_data[col] = le.fit_transform(categorical_data[col])
    # Apply LabelEncoder to each column separately in a loop

In [43]:
categorical_data

workclass    education       marital_status  \
0              State-gov    Bachelors        Never-married   
1       Self-emp-not-inc    Bachelors   Married-civ-spouse   
2                Private      HS-grad             Divorced   
3                Private         11th   Married-civ-spouse   
4                Private    Bachelors   Married-civ-spouse   
...                  ...          ...                  ...   
32556            Private   Assoc-acdm   Married-civ-spouse   
32557            Private      HS-grad   Married-civ-spouse   
32558            Private      HS-grad              Widowed   
32559            Private      HS-grad        Never-married   
32560       Self-emp-inc      HS-grad   Married-civ-spouse   

               occupation    relationship  race  sex  native_country  income  
0            Adm-clerical   Not-in-family     4    1   United-States       0  
1         Exec-managerial         Husband     4    1   United-States       0  
2       Handlers-cleaners   Not-in-family     4    1   United-States       0  
3       Handlers-cleaners         Husband     2    1   United-States       0  
4          Prof-specialty            Wife     2    0            Cuba       0  
...                   ...             ...   ...  ...             ...     ...  
32556        Tech-support            Wife     4    0   United-States       0  
32557   Machine-op-inspct         Husband     4    1   United-States       1  
32558        Adm-clerical       Unmarried     4    0   United-States       0  
32559        Adm-clerical       Own-child     4    1   United-States       0  
32560     Exec-managerial            Wife     4    0   United-States       1  

[32561 rows x 9 columns]

**One Hot Encoding**

In [44]:
categorical_data = pd.get_dummies(categorical_data, columns=['workclass', 'education', 'marital_status', 'occupation', 'relationship', 'native_country'])

# Convert True/False to 1/0
categorical_data = categorical_data.astype(int)
categorical_data

race  sex  income  workclass_ ?  workclass_ Federal-gov  \
0         4    1       0             0                       0   
1         4    1       0             0                       0   
2         4    1       0             0                       0   
3         2    1       0             0                       0   
4         2    0       0             0                       0   
...     ...  ...     ...           ...                     ...   
32556     4    0       0             0                       0   
32557     4    1       1             0                       0   
32558     4    0       0             0                       0   
32559     4    1       0             0                       0   
32560     4    0       1             0                       0   

       workclass_ Local-gov  workclass_ Never-worked  workclass_ Private  \
0                         0                        0                   0   
1                         0                        0                   0   
2                         0                        0                   1   
3                         0                        0                   1   
4                         0                        0                   1   
...                     ...                      ...                 ...   
32556                     0                        0                   1   
32557                     0                        0                   1   
32558                     0                        0                   1   
32559                     0                        0                   1   
32560                     0                        0                   0   

       workclass_ Self-emp-inc  workclass_ Self-emp-not-inc  ...  \
0                            0                            0  ...   
1                            0                            1  ...   
2                            0                            0  ...   
3                            0                            0  ...   
4                            0                            0  ...   
...                        ...                          ...  ...   
32556                        0                            0  ...   
32557                        0                            0  ...   
32558                        0                            0  ...   
32559                        0                            0  ...   
32560                        1                            0  ...   

       native_country_ Portugal  native_country_ Puerto-Rico  \
0                             0                            0   
1                             0                            0   
2                             0                            0   
3                             0                            0   
4                             0                            0   
...                         ...                          ...   
32556                         0                            0   
32557                         0                            0   
32558                         0                            0   
32559                         0                            0   
32560                         0                            0   

       native_country_ Scotland  native_country_ South  \
0                             0                      0   
1                             0                      0   
2                             0                      0   
3                             0                      0   
4                             0                      0   
...                         ...                    ...   
32556                         0                      0   
32557                         0                      0   
32558                         0                      0   
32559                         0                      0   
32560                         0                      0   

       native_coun

**Discuss the pros and cons of One-Hot Encoding and Label Encoding.**

# One-Hot Encoding

One-Hot Encoding (OHE) is a technique that converts categorical variables into a series of binary variables (0s and 1s). Each category is represented by a separate column.

***Pros:***
1. No Ordinal Relationship Assumed: One-Hot Encoding does not assume any inherent order between the categories, making it ideal for nominal (unordered) categorical variables.

2. Machine Learning Compatibility: Many machine learning algorithms, especially those that rely on distance metrics (e.g., K-Nearest Neighbors, SVMs), work better with OHE because it avoids assigning arbitrary relationships between categories.

3. Interpretability: Each column in OHE is easily interpretable as representing a specific category. This can make the output of models easier to understand.

***Cons:***

1. High Dimensionality: For categorical variables with many unique categories, OHE can create a large number of columns, leading to the "curse of dimensionality." This can increase the complexity of the model and require more computational resources.

2. Sparse Data: Since most of the values in the resulting matrix are 0, OHE creates sparse data, which can be inefficient to store and process.

3. Potential for Overfitting: High-dimensional feature spaces created by OHE can lead to overfitting, especially with small datasets.

# Label Encoding
Label Encoding converts each category into a unique integer based on its label. It assigns a numerical value to each category, typically in the order in which they appear.

***Pros:***

1. Simplicity: Label Encoding is straightforward to implement and results in a single column, keeping the dimensionality low.

2. Efficient Representation: Since it doesn't expand the number of columns, Label Encoding is memory-efficient and can be faster to process, especially with large datasets.

3. Useful for Ordinal Data: When the categories have a meaningful order (e.g., "Low," "Medium," "High"), Label Encoding can preserve this ordinal relationship.

***Cons:***

1. Assumes Ordinal Relationship: Label Encoding imposes an ordinal relationship between categories, even if none exists. This can be problematic for nominal data where no such relationship should be assumed.

2. Can Mislead Algorithms: Some machine learning algorithms may interpret the numerical values as having a natural order or distance between them, which can lead to biased or incorrect results (e.g., treating "1" and "2" as closer than "1" and "3").

3. Poor Performance with Some Algorithms: Algorithms that rely on distance metrics or relationships between features (like KNN or SVM) may not perform well with Label Encoding, as the numerical values may introduce unintended biases.

# Feature Engineering

In [45]:
# Creating new feature 'age_group'
bins = [0, 18, 30, 45, 60, 100]
labels = ['Teen', 'Young Adult', 'Adult', 'Middle-Aged', 'Senior']
data['age_group'] = pd.cut(data['age'], bins=bins, labels=labels, right=False)

# Creating new feature 'hours_per_week_category'
bins = [0, 20, 40, 60, 100]
labels = ['Part-Time', 'Full-Time', 'Overtime', 'Workaholic']
data['hours_per_week_category'] = pd.cut(data['hours_per_week'], bins=bins, labels=labels, right=False)


In [46]:
data

age          workclass  fnlwgt    education  education_num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital_gain  capital_loss  hours_per_week  native_country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

       income    age_group hours_per_week_category  
0       <=50K        Adult                Overtime  
1       <=50K  Middle-Aged               Part-Time  
2       <=50K        Adult                Overtime  
3       <=50K  Middle-Aged                Overtime  
4       <=50K  Young Adult                Overtime  
...       ...          ...                     ...  
32556   <=50K  Young Adult               Full-Time  
32557    >50K        Adult                Overtime  
32558   <=50K  Middle-Aged                Overtime  
32559   <=50K  Young Adult               Full-Time  
32560    >50K  Middle-Aged                Overtime  

[32561 rows x 17 columns]

In [47]:
numerical_data.skew()

age                0.558743
fnlwgt             1.446980
education_num     -0.311676
capital_gain      11.953848
capital_loss       4.594629
hours_per_week     0.227643
dtype: float64

**capital_gain has too much positive skewness, lets log transform it.**

In [48]:
# Apply log transformation
numerical_data['capital_gain'] = np.log(numerical_data['capital_gain'])

/usr/local/lib/python3.10/dist-packages/pandas/core/arraylike.py:402: RuntimeWarning: invalid value encountered in log
  result = getattr(ufunc, method)(*inputs, **kwargs)


In [49]:
numerical_data.skew()

age               0.558743
fnlwgt            1.446980
education_num    -0.311676
capital_gain      0.076800
capital_loss      4.594629
hours_per_week    0.227643
dtype: float64

# Feature Selection

In [50]:
data = pd.concat([numerical_data, categorical_data], axis=1)
data

age    fnlwgt  education_num  capital_gain  capital_loss  \
0      0.030671 -1.063611       1.134739     -1.907488      -0.21666   
1      0.837109 -1.008707       1.134739           NaN      -0.21666   
2     -0.042642  0.245079      -0.420060           NaN      -0.21666   
3      1.057047  0.425801      -1.197459           NaN      -0.21666   
4     -0.775768  1.408176       1.134739           NaN      -0.21666   
...         ...       ...            ...           ...           ...   
32556 -0.849080  0.639741       0.746039           NaN      -0.21666   
32557  0.103983 -0.335433      -0.420060           NaN      -0.21666   
32558  1.423610 -0.358777      -0.420060           NaN      -0.21666   
32559 -1.215643  0.110960      -0.420060           NaN      -0.21666   
32560  0.983734  0.929893      -0.420060      0.635743      -0.21666   

       hours_per_week  race  sex  income  workclass_ ?  ...  \
0           -0.035429     4    1       0             0  ...   
1           -2.222153     4    1       0             0  ...   
2           -0.035429     4    1       0             0  ...   
3           -0.035429     2    1       0             0  ...   
4           -0.035429     2    0       0             0  ...   
...               ...   ...  ...     ...           ...  ...   
32556       -0.197409     4    0       0             0  ...   
32557       -0.035429     4    1       1             0  ...   
32558       -0.035429     4    0       0             0  ...   
32559       -1.655225     4    1       0             0  ...   
32560       -0.035429     4    0       1             0  ...   

       native_country_ Portugal  native_country_ Puerto-Rico  \
0                             0                            0   
1                             0                            0   
2                             0                            0   
3                             0                            0   
4                             0                            0   
...                         ...                          ...   
32556                         0                            0   
32557                         0                            0   
32558                         0                            0   
32559                         0                            0   
32560                         0                            0   

       native_country_ Scotland  native_country_ South  \
0                             0                      0   
1                             0                      0   
2                             0                      0   
3                             0                      0   
4                             0                      0   
...                         ...                    ...   
32556                         0                      0   
32557                         0                      0   
32558                         0                      0   
32559                         0                      0   
32560                         0                      0   

       native_country_ Taiwan  native_country_ Thailand  \
0                           0                         0   
1                           0                         0   
2                           0                         0   
3                           0                         0   
4                           0                         0   
...                       ...                       ...   
32556                       0                         0   
32557                       0                         0   
32558                       0                         0   
32559                       0                         0   
32560                       0                         0   

       native_country_ Trinadad&Tobago  native_country_ United-States  \
0                                    0                              1   
1                                    0                              1

In [51]:
data.isnull().sum()

age                                    0
fnlwgt                                 0
education_num                          0
capital_gain                       29929
capital_loss                           0
                                   ...  
native_country_ Thailand               0
native_country_ Trinadad&Tobago        0
native_country_ United-States          0
native_country_ Vietnam                0
native_country_ Yugoslavia             0
Length: 104, dtype: int64

In [52]:
data.drop('capital_gain', axis=1, inplace=True)

**Isolation Forest algorithm**

In [53]:
from sklearn.ensemble import IsolationForest

# Initialize the Isolation Forest model
iso_forest = IsolationForest(contamination=0.1, random_state=42)

# Fit the model
iso_forest.fit(data)

IsolationForest(contamination=0.1, random_state=42)

In [54]:
# Predict the outliers (-1 indicates an outlier, 1 indicates an inlier)
data['Anomaly'] = iso_forest.predict(data)

# Identify and remove outliers
df_inliers = data[data['Anomaly'] == 1]
df_outliers = data[data['Anomaly'] == -1]

# Drop the 'Anomaly' column if it's no longer needed
df_inliers = df_inliers.drop(columns=['Anomaly'])

In [55]:
len(df_outliers)

3256

**There are 3256 total outliers.**

***Outliers can negatively affect model performance in the following key ways:***

**1. Bias and Skew in Statistical Models:** In models like linear regression, outliers can disproportionately influence the model, leading to biased results and poor predictions.

**2. Misclassification in Distance-Based Models:** Algorithms like KNN and K-Means are sensitive to outliers, which can distort distance calculations and cluster assignments, reducing overall accuracy.

**3. Evaluation Metric Distortion:** Outliers can inflate error metrics like Mean Squared Error (MSE), leading to a misleading assessment of model performance.

In [56]:
!pip install ppscore

In [29]:

# prompt: calculate pps score for finding the relationship between columns

import ppscore as pps

# Calculate PPS matrix
pps_matrix = pps.matrix(df_inliers)

# Display the PPS matrix
pps_matrix

x                                y  ppscore  \
0                             age                              age      1.0   
1                             age                           fnlwgt      0.0   
2                             age                    education_num      0.0   
3                             age                     capital_loss      0.0   
4                             age                   hours_per_week      0.0   
...                           ...                              ...      ...   
10604  native_country_ Yugoslavia         native_country_ Thailand      0.0   
10605  native_country_ Yugoslavia  native_country_ Trinadad&Tobago      0.0   
10606  native_country_ Yugoslavia    native_country_ United-States      0.0   
10607  native_country_ Yugoslavia          native_country_ Vietnam      0.0   
10608  native_country_ Yugoslavia       native_country_ Yugoslavia      1.0   

                     case  is_valid_score               metric  \
0          predict_itself            True                 None   
1              regression            True  mean absolute error   
2              regression            True  mean absolute error   
3              regression            True  mean absolute error   
4              regression            True  mean absolute error   
...                   ...             ...                  ...   
10604          regression            True  mean absolute error   
10605  target_is_constant            True                 None   
10606          regression            True  mean absolute error   
10607          regression            True  mean absolute error   
10608      predict_itself            True                 None   

       baseline_score  model_score                    model  
0            0.000000     1.000000                     None  
1            0.718557     0.729824  DecisionTreeRegressor()  
2            0.676104     0.692483  DecisionTreeRegressor()  
3            0.182332     0.344261  DecisionTreeRegressor()  
4            0.579822     0.616957  DecisionTreeRegressor()  
...               ...          ...                      ...  
10604        0.000400     0.000800  DecisionTreeRegressor()  
10605        1.000000     1.000000                     None  
10606        0.077400     0.141842  DecisionTreeRegressor()  
10607        0.001600     0.003195  DecisionTreeRegressor()  
10608        0.000000     1.000000                     None  

[10609 rows x 9 columns]

A PPS of,
* 0 means no predictive power
* PPS closer to 1 indicates strong predictive power.